src/main/java/org/opensearch/security/http/HTTPOnBehalfOfJwtAuthenticator.java

```diff
@@ -0,0 +1,277 @@
+            throw new RuntimeException(e);
+        }
+
+        subjectKey = "sub";
+    }
+
+    @Override
+    @SuppressWarnings("removal")
+    public AuthCredentials extractCredentials(RestRequest request, ThreadContext context) throws OpenSearchSecurityException {
+        final SecurityManager sm = System.getSecurityManager();
+
+        if (sm != null) {
+            sm.checkPermission(new SpecialPermission());
+        }
+
+        AuthCredentials creds = AccessController.doPrivileged(new PrivilegedAction<AuthCredentials>() {
+            @Override
+            public AuthCredentials run() {
+                return extractCredentials0(request);
+            }
+        });
+
+        return creds;
+    }
+
+    private AuthCredentials extractCredentials0(final RestRequest request) {
+        if (jwtParser == null) {
+            log.error("Missing Signing Key. JWT authentication will not work");
+            return null;
+        }
+
+        String jwtToken = request.header(HttpHeaders.AUTHORIZATION);
+
+        if (jwtToken == null || jwtToken.length() == 0) {
+            if(log.isDebugEnabled()) {
+                log.debug("No JWT token found in '{}' header", HttpHeaders.AUTHORIZATION);
+            }
+            return null;
+        }
+
+        if (!BEARER.matcher(jwtToken).matches()) {
+            jwtToken = null;
+        }
+
+        final int index;
+        if((index = jwtToken.toLowerCase().indexOf(BEARER_PREFIX)) > -1) { //detect Bearer
+            jwtToken = jwtToken.substring(index+BEARER_PREFIX.length());
+        } else {
+            if(log.isDebugEnabled()) {
+                log.debug("No Bearer scheme found in header");
+            }
+        }
+
+        try {
+            final Claims claims = jwtParser.parseClaimsJws(jwtToken).getBody();
+
+            final String subject = extractSubject(claims, request);
+
+            final String audience = claims.getAudience();
+
+            //TODO: GET ROLESCLAIM DEPENDING ON THE STATUS OF BWC MODE. ON: er / OFF: dr
+            Object rolesObject = null;
+            String[] roles;
+
+            try {
+                rolesObject = claims.get("er");
+            } catch (Throwable e) {
+                    log.debug("No encrypted role founded in the claim, continue searching for decrypted roles.");
+            }
+
+            try {
+                rolesObject = claims.get("dr");
+            } catch (Throwable e) {
+                log.debug("No decrypted role founded in the claim.");
+            }
+
+            if (rolesObject == null) {
+                log.warn(
+                        "Failed to get roles from JWT claims. Check if this key is correct and available in the JWT payload.");
+                roles = new String[0];
+            } else {
+                final String rolesClaim = rolesObject.toString();
+
+                // Extracting roles based on the compatbility mode
+                String rolesPostEncryptionDecryption = rolesClaim;
```

Comment:
nit: Can we change this variable name to `decryptedRoles`?

Reply:
Sure, I have fixed.